src/ReservoirRouter.sol

```diff
@@ -0,0 +1,291 @@
+            }
+            else {
+                amount = output == token0 ? int256(amounts[i + 1]) : -int256(amounts[i + 1]);
+            }
+
+            IReservoirPair(ReservoirLibrary.pairFor(address(factory), input, output, curveIds[i])).swap(
+                amount, inOrOut, to, new bytes(0)
+            );
+
+            unchecked { i += 1; }
+        }
+    }
+
+    function swapExactForVariable(
+        uint256 amountIn,
+        uint256 amountOutMin,
+        address[] calldata path,
+        uint256[] calldata curveIds,
+        address to
+    ) external payable returns (uint256[] memory amounts) {
+        amounts = ReservoirLibrary.getAmountsOut(address(factory), amountIn, path, curveIds);
+        // but the actual swap results might be diff from this. Should we move the require into _swap to check for the minOut?
+        require(amounts[amounts.length - 1] >= amountOutMin, "RL: INSUFFICIENT_OUTPUT_AMOUNT");
+
+        _pay(path[0], msg.sender, ReservoirLibrary.pairFor(address(factory), path[0], path[1], curveIds[0]), amounts[0]);
+        _swap(amounts, true, path, curveIds, to);
+    }
+
+    function swapVariableForExact(
+        uint256 amountOut,
+        uint256 amountInMax,
+        address[] calldata path,
+        uint256[] calldata curveIds,
+        address to
+    ) external payable returns (uint256[] memory amounts) {
+        amounts = ReservoirLibrary.getAmountsIn(address(factory), amountOut, path, curveIds);
+        require(amounts[0] <= amountInMax, "RL: EXCESSIVE_INPUT_AMOUNT");
+
+        _pay(path[0], msg.sender, ReservoirLibrary.pairFor(address(factory), path[0], path[1], curveIds[0]), amounts[0]);
+        _swap(amounts, false, path, curveIds, to);
+    }
+
+    function getAmountOut(
+        uint256 amountIn,
+        uint256 reserveIn,
+        uint256 reserveOut,
+        uint256 curveId,
+        uint256 swapFee,
+        ExtraData calldata extraData
+    ) external pure returns (uint256 amountOut) {
+        if (curveId == 0) {
+            return ReservoirLibrary.getAmountOutConstantProduct(amountIn, reserveIn, reserveOut, swapFee);
+        }
+        else if (curveId == 1) {
+            return ReservoirLibrary.getAmountOutStable(amountIn, reserveIn, reserveOut, swapFee, extraData);
+        }
+    }
+
+    function getAmountIn(
+        uint256 amountOut,
+        uint256 reserveIn,
+        uint256 reserveOut,
+        uint256 curveId,
+        uint256 swapFee,
+        ExtraData calldata extraData
+    ) external pure returns (uint256 amountIn) {
+        if (curveId == 0) {
+            return ReservoirLibrary.getAmountInConstantProduct(amountOut, reserveIn, reserveOut, swapFee);
+        }
+        else if (curveId == 1) {
+            return ReservoirLibrary.getAmountInStable(amountOut, reserveIn, reserveOut, swapFee, extraData);
+        }
+    }
+
+    // perf: to use calldata or memory for path and curveIds?
```

Comment:
Calldata, only use memory if you must call it from within the contract. Because this is `external` using `calldata` will prevent a memory allocation unless you need it

Reply:
Especially because the library uses calldata too, it should totally avoid the memory allocation. I can test this out when we switch to optimization mode.

Reply:
okay removed comment